proposals/p1964.md

```diff
@@ -0,0 +1,140 @@
+
+## Problem
+
+This proposal specifies lexical rules for constant characters in Carbon.
+
+## Background
+
+We wish to provide a distinct lexical syntax for character literals versus
+string literals.
+
+In theory we could just reuse string literals for the purpose of character
+literals. However, it could benefit the readablity of our code if we had a
+distinct lexical syntax for character literals versus string literals.
+
+## Proposal
+
+The idea is to create and manage a character literal the same we would as a
+string, but using the single quote delimiter (') compared to the string double
+quote (").
+
+As with string literals, each character literal would have a different type.
+
+    var w: Character = 'w';
+
+We will not support:
+
+-   Multi-line literals
+-   "raw" literals (using #'x'#)
+-   Empty character literals (''')
+
+## Details
+
+A character literal is a sequence enclosed with single quotes delimiter ('),
+excluding:
+
+-   New line
+-   Single quote (`'`)
+-   Back-slash (`\`)
+-   Escape sequences
+
+The type of a character literal will depend on the the contents, so that `'c'`
+and `'b'`. However any `'\n'` and `'\u{A}'` would be of the same type (As when
+they are encoded, they are the same unicode entities `%0A`).
+
+### Types
+
+The different character literal types are based on the value of the character.
+We will not be supporting prefix declarations like `u`, `U`, or `L`. The only
+difference between character types is the type of variable you assign the
+character to. There is no explicit difference between 8-bit and 16-bit
+characters when declaring the literal.
+
+### Encoding
+
+They type of character literal and the way it is encoded should directly depends
+on the type of variable you assign it to, and these different types should be
+encoded as follows:
+
+-   Ordinary (UTF-8) character literals should use a single UTF-8 code unit.
+-   Wide-character literals should use single Unicode code point.
+-   UTF-16 character literals should use a single Unicode code point.
+-   UTF-32 character literals should use a single Unicode code point.
+-   Glyph character literals should use a base character (Single Unicode point)
+    plus a sequence of combining characters.
+
+This is experimental, and should be revisited if we find motivation for
+expressing character literals in other encodings.
+
+## Rationale
+
+This proposal supports the goal of making Carbon code
+[easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+and
+[Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+by ensuring that every kind of character literal that exists in C++ can be
+represented in a Carbon character literal. This is done in a way that is natural
+to adopt, understand, easy to read by having explicit character types mapped to
+the C++ character types and the correct associated encoding.
```

Comment:
This rationale supports the C++ interoperability goal. Could you have something else about why this makes code easy to read, understand, and write? You could possibly talk about it being concise, familiar, and/or consistent with other things in Carbon (like numeric literals or string literals).